src/main/java/ch/njol/skript/effects/EffToggle.java

```diff
@@ -40,63 +41,84 @@
+			"(toggle|switch) [[the] state of] %booleans/blocks%"
+			);
+	}
+
+	private enum State {
+    ACTIVATE, DEACTIVATE, TOGGLE
 	}
 
 	@SuppressWarnings("null")
-	private Expression<Block> blocks;
-	private int toggle;
+	private Expression<?> toggledExpr;
+	private State state;
 	
-	@SuppressWarnings({"unchecked", "null"})
 	@Override
-	public boolean init(final Expression<?>[] vars, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
-		blocks = (Expression<Block>) vars[0];
-		toggle = matchedPattern - 1;
+	public boolean init(Expression<?>[] vars, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		toggledExpr = (Expression<?>) vars[0];
+		state = State.values()[matchedPattern];
+		if (toggledExpr.getReturnType() == Boolean.class && !ChangerUtils.acceptsChange(toggledExpr, ChangeMode.SET, Boolean.class)) {
+			Skript.error(toggledExpr.toString(null, false) + " cannot be toggled");
+			return false;
+		}
 		return true;
 	}
 
 	@Override
-	protected void execute(final Event e) {
-		for (Block b : blocks.getArray(e)) {
-			BlockData data = b.getBlockData();
-			if (toggle == -1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(false);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(false);
-			} else if (toggle == 1) {
-				if (data instanceof Openable)
-					((Openable) data).setOpen(true);
-				else if (data instanceof Powerable)
-					((Powerable) data).setPowered(true);
-			} else {
-				if (data instanceof Openable) // open = NOT was open
-					((Openable) data).setOpen(!((Openable) data).isOpen());
-				else if (data instanceof Powerable) // power = NOT power
-					((Powerable) data).setPowered(!((Powerable) data).isPowered());
+	protected void execute(Event event) {
+		ArrayList<Object> toggledValues = new ArrayList<>();
+		for (Object obj : toggledExpr.getArray(event)) {
+			if (obj instanceof Block) {
+				Block block = (Block) obj;
+				BlockData data = block.getBlockData();
+				if (state == State.TOGGLE) {
+					if (data instanceof Openable) // open = NOT was open
+						((Openable) data).setOpen(!((Openable) data).isOpen());
+					else if (data instanceof Powerable) // power = NOT power
+						((Powerable) data).setPowered(!((Powerable) data).isPowered());
+				} else {
+					boolean value = state == State.ACTIVATE; 
+					if (data instanceof Openable)
+						((Openable) data).setOpen(value);
+					else if (data instanceof Powerable)
+						((Powerable) data).setPowered(value);
+				}
+
+				block.setBlockData(data);
+				toggledValues.add(block);
+
+			} else if (obj instanceof Boolean) {
+				toggledValues.add(!(Boolean) obj);
 			}
-			
-			b.setBlockData(data);
 		}
-	}
 
+		List<Object> filteredValues = toggledValues.stream().filter(
+			obj -> ChangerUtils.acceptsChange(toggledExpr, ChangeMode.SET, obj.getClass())
+		).collect(Collectors.toList());
+		toggledExpr.change(event, filteredValues.toArray(), ChangeMode.SET);
```

Comment:
```suggestion
		Object[] filteredValues = toggledValues.stream().filter(
			obj -> ChangerUtils.acceptsChange(toggledExpr, ChangeMode.SET, obj.getClass())
		).toArray(Object[]::new);
		toggledExpr.change(event, filteredValues, ChangeMode.SET);
```
You should be able to collect the stream elements directly into an array.